src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -1091,12 +1091,15 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           @inline def tpdPos(transformed: Tree) = typedPos(tree.pos, mode, pt)(transformed)
           @inline def tpd(transformed: Tree)    = typed(transformed, mode, pt)
 
-          @inline def warnValueDiscard(): Unit = if (!isPastTyper && settings.warnValueDiscard) {
```

Comment:
I would prefer to keep this new warning under the flag as well.

Reply:
I think this was intended to have the moral force of deprecating `Unit` value. I don't think we considered actually issuing it as a deprecation, which as someone recently discussed, really does just mean, Don't do that.

Reply:
Brilliant! Why not give `Unit` (the companion) the actual force of `@deprecate` to tersely point out it's compiler internal, and maybe you meant `()`? Sounds like a win-win to me. (There's plenty more of internal compiler stuff that should not be used from user-level, actually.)

Reply:
`Unit` object can be legitimately used as a parameter to `@specialized` annotation, so I don't think it can just be deprecated. We could of course also adjust the annotation, but introducing a special case there just to avoid a special case here doesn't seem like a clear win to me.